### PR TITLE
Easy speed up using reciprocals.

### PR DESCRIPTION
**Pull request by proteneer**
This optimizes the norm calculation for both real and complex distance functions to use rnorm variants (which have only 1/2 ULPS of precision loss in double and single precision respectively).

### CODE REVIEW

File: timemachine/cpp/src/kernels/k_nonbonded_deterministic.cuh [hunk 1]
```diff
@@ -114,26 +182,28 @@ void __global__ k_nonbonded_jvp(
     Surreal<RealType> g_sigj(0.0, 0.0);
     Surreal<RealType> g_epsj(0.0, 0.0);
 
+    RealType inv_cutoff = 1/cutoff;
```
Comment (bp-kelley) [inline timemachine/cpp/src/kernels/k_nonbonded_deterministic.cuh]: I would use Static_cast<RealType> 1.0/cutoff if just to avoid compiler earnings.

Comment (bp-kelley) [reply timemachine/cpp/src/kernels/k_nonbonded_deterministic.cuh]: If you want to be really fancy see rd_cast in the rdkit sources.  In debug mode it throws an error on overflow.

File: timemachine/cpp/src/kernels/k_nonbonded_deterministic.cuh [hunk 2]
```diff
@@ -275,21 +345,23 @@ void __global__ k_nonbonded_inference(
     RealType sig_j = atom_j_idx < N ? params[lj_param_idx_sig_j] : 1;
     RealType eps_j = atom_j_idx < N ? params[lj_param_idx_eps_j] : 0;
 
+    RealType inv_cutoff = 1/cutoff;
+
     // In inference mode, we don't care about gradients with respect to parameters.
     for(int round = 0; round < 32; round++) {
```
Comment (bp-kelley) [inline timemachine/cpp/src/kernels/k_nonbonded_deterministic.cuh]: 32 looks magic, perhaps use a named constant?

**Comment by proteneer** [comment]
We can get a bit more of a speed up (another 12%) if we don't differentiate with respect to the vdw parameters - but I think that's a little too aggressive for now so I'm going to keep this as is.

File: timemachine/cpp/src/kernels/k_nonbonded_deterministic.cuh (hunk 1 above)
Comment (proteneer) [reply timemachine/cpp/src/kernels/k_nonbonded_deterministic.cuh]: I'm using the templated OtherType method we spoke about so there shouldn't be any warnings.

File: timemachine/cpp/src/kernels/k_nonbonded_deterministic.cuh (hunk 2 above)
Comment (proteneer) [reply timemachine/cpp/src/kernels/k_nonbonded_deterministic.cuh]: 32 is WARPSIZE - we will have a lot of problems on our hands if this is ever modified to be not 32

**Pull request merged**